scrapy/spidermiddlewares/offsite.py

```diff
@@ -42,7 +42,8 @@ def process_spider_output(self, response, result, spider):
                 yield x
 
     def should_follow(self, request, spider):
-        regex = self.host_regex
+        # Possibly update allowed_domains.
+        regex = self.get_host_regex(spider)
```

Comment:
hey! I wonder what's performance overhead of doing it on each request, especially for large allowed_domains lists. Do you think we can have some cache, to avoid building regex if allowed_domains hasn't changed?

Reply:
I’m hoping that @eliasdorneles’s https://github.com/scrapy/scrapy/pull/1640#issuecomment-182451216 stands. Although it only caches 100 patterns, as far as I know, so maybe a custom cache makes sense.

Reply:
That's right, though regex compiling is not the only work which is being done. Maybe it makes sense to do a micro-benchmark, e.g. check how much time does it take to do get_host_regex for 1K allowed_domains.

Reply:
Use case I'm thinking about: start_requests for 4K seeds; allowed_domains is populated to limit crawl only to those websites; spider is doing minimal processing (e.g. discovering links). In this case we can expect a spider to perform at ~300-600 pages per minute, maybe 1000 pages per minute; I wonder how these numbes change after this PR.

Reply:
```py
import re
import timeit
import warnings

class MockSpider:
    allowed_domains = ['example{}.com'.format(i) for i in range(1000)]


spider = MockSpider()


def get_host_regex(spider):
    """Override this method to implement a different offsite policy"""
    allowed_domains = getattr(spider, 'allowed_domains', None)
    if not allowed_domains:
        return re.compile('')  # allow all by default
    url_pattern = re.compile(r"^https?://.*$")
    port_pattern = re.compile(r":\d+$")
    domains = []
    for domain in allowed_domains:
        if domain is None:
            continue
        elif url_pattern.match(domain):
            message = ("allowed_domains accepts only domains, not URLs. "
                        "Ignoring URL entry %s in allowed_domains." % domain)
            warnings.warn(message, URLWarning)
        elif port_pattern.search(domain):
            message = ("allowed_domains accepts only domains without ports. "
                        "Ignoring entry %s in allowed_domains." % domain)
            warnings.warn(message, PortWarning)
        else:
            domains.append(re.escape(domain))
    regex = r'^(.*\.)?(%s)$' % '|'.join(domains)
    return re.compile(regex)

def _old():
    get_host_regex(spider)

for n in (1, 1, 10, 100, 1000, 10000):
    print('{}: {}'.format(n, timeit.timeit(_old, number=n)))
```

```
$ python test_allowed_domains_regex.py 
1: 0.052881918993080035
1: 0.003914863002137281
10: 0.03774952599633252
100: 0.378773693999392
1000: 3.7923534260044107
10000: 38.02031098899897
```

Reply:
Thanks @Gallaecio for the benchmarking it! Do I read the number right, is it 3.8s of CPU time per request if there is 1K domains in allowed_domains, 38s if there are 10K, and 0.37s for the humble 100 domains?

Reply:
No, the number of domains is fixed at 1k, that is the number of calls.

Reply:
So, 38s for 10,000 requests with 1,000 allowed domains.